docs for 1.2 relnotes (#500)
agentgateway/website: content/docs/kubernetes/main/reference/release-notes.md

For more details, review the [GitHub release notes in the agentgateway repository](https://github.com/agentgateway/agentgateway/releases)
{{< /callout >}}

## 🔥 Breaking changes {#v12-breaking-changes}

### xDS Helm configuration changed

The controller Helm value for xDS TLS changed from a boolean to an explicit transport mode.

Before:

```yaml
controller:
  xds:
    tls:
      enabled: true
```

Now:

```yaml
controller:
  xds:
    mode: tls
```

> Supported values are `plaintext`, `tls`, and `either`. The chart now defaults to `tls`, and the controller can automatically manage xDS TLS material. If you have automation that sets `controller.xds.tls.enabled`, update it to use `controller.xds.mode`. For more information, see [TLS encryption]({{< link-hextra path="/install/tls/" >}}).

Reviewer: ```suggestion
Supported values are `plaintext`, `tls`, and `either`. The chart now defaults to `tls` to enable TLS encryption for the controller. If you have automation that sets `controller.xds.tls.enabled`, update it to use `controller.xds.mode`. For more information, see [TLS encryption]({{< link-hextra path="/install/tls/" >}}).
```

spec:
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /anything/team1/foo
    backendRefs:
    - name: httpbin
      port: 8000
```

### `agctl` CLI debugging tool

A new experimental `agctl` command-line tool for inspecting and debugging agentgateway is now available. To install `agctl`, see [Install agctl]({{< link-hextra path="/operations/agctl/" >}}).
- `agctl config` renders the runtime configuration that an agentgateway proxy has loaded, including binds, listeners, routes, backends, workloads, and policies, as a structured table, JSON, or YAML. `agctl config backends` shows per-backend health, request counts, and latency. For more information, see [Inspect agentgateway configuration]({{< link-hextra path="/operations/inspect-config/" >}}).
- `agctl trace` streams a step-by-step trace of how the proxy processes the next request, showing the matched route, applied policies, chosen backend, and response status. For more information, see [Trace requests with agctl]({{< link-hextra path="/operations/trace-requests/" >}}).

For a complete command reference, see [agctl CLI reference]({{< link-hextra path="/reference/agctl/" >}}).

### Policy targets

> `AgentgatewayPolicy` can now target resources by label selector in addition to explicit target references. This makes it easier to apply shared policy configuration across groups of Gateways, Routes, Services, or Backends.

Reviewer: ```suggestion
The `AgentgatewayPolicy` resource can now target resources by label selector in addition to explicit target references. This configuration makes it easier to apply shared policy configuration across groups of Gateways, Routes, Services, or Backends.
```

> Additionally, `AgentgatewayPolicy` can now target `ListenerSet` and `InferencePool` resources. For more information, see [Targeting and merging]({{< link-hextra path="/about/policies/target-merge/" >}}).

Reviewer: ```suggestion
Additionally, the `AgentgatewayPolicy` resource can now target `ListenerSet` and `InferencePool` resources. For more information, see [Targeting and merging]({{< link-hextra path="/about/policies/target-merge/" >}}).
```

- **Istio workload TLS**: Listeners can use Istio workload certificates for simple TLS or mutual TLS. For more information, see [mTLS]({{< link-hextra path="/setup/listeners/mtls/" >}}).
- **HBONE gateway tunnel protocol**: Controller support was added for the `HBONE_GATEWAY` tunnel protocol.
- **Unix Domain Socket backends**: Kubernetes static backends can now target Unix Domain Sockets.
- **Max connection duration**: HTTP listeners can now enforce a maximum connection duration.
- **HTTP/2 pooling**: HTTP/2 connection pooling is improved to avoid the single-connection bottleneck.

### Operations

- Agentgateway's memory allocator performance is improved, resulting in increased runtime performance and decreased memory utilization.
- A new `/debug/pprof/heap` endpoint is available to get a `pprof` snapshot of current and historical allocations.

### Telemetry

- **Custom Prometheus labels**: `AgentgatewayPolicy` can add custom Prometheus metric labels using CEL expressions.

> - **OpenTelemetry environment variables**: OTEL configuration now respects standard environment variables. For more information, see [OTel stack]({{< link-hextra path="/observability/otel-stack/" >}}).

Reviewer: ```suggestion
- **OpenTelemetry environment variables**: The OTEL configuration now respects standard environment variables. For more information, see [OTel stack]({{< link-hextra path="/observability/otel-stack/" >}}).
```

## 🪲 Notable fixes {#v12-fixes}

- Fixed A2A policy matching for agents hosted under sub-paths.
- Fixed A2A and MCP handling of `X-Forwarded-Proto`.
- Fixed service parents and arbitrary parents in route delegation.

> - Fixed phantom backend chains attaching policies to missing targets.

Author: ```suggestion
- Fixed policies that target missing or non-existent backends from silently attaching to the gateway.
```
